### ISSUE TITLE
wsj_mix cracks in min_pit_process if report_cer and report_wer set to be True

### ISSUE THREAD

**Issue by vyouman** [espnet/espnet#633]
I want to do some experiments on WSJ0_mix, but it seems that it cracks at the end of an epoch.
Here is the exception in the main training loop. 

> ^[[JException in main training loop: list index out of range
Traceback (most recent call last):
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/chainer/training/trainer.py", line 318, in run
    entry.extension(self)
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/chainer/training/extensions/evaluator.py", line 157, in __call__
    result = self.evaluate()
  File "/home/user/vyouman/espnet/espnet/asr/pytorch_backend/asr.py", line 104, in evaluate
    self.model(*x)
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 477, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py", line 123, in forward
    outputs = self.parallel_apply(replicas, inputs, kwargs)
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py", line 133, in parallel_apply
    return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/parallel/parallel_apply.py", line 77, in parallel_apply
    raise output
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/parallel/parallel_apply.py", line 53, in _worker
    output = module(*input, **kwargs)
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 477, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/user/vyouman/espnet/espnet/nets/pytorch_backend/e2e_asr_mix.py", line 303, in forward
    wers.append(self.min_pit_process(tmp_wers) / len(sum(ref_words, [])))
  File "/home/user/vyouman/espnet/espnet/nets/pytorch_backend/e2e_asr_mix.py", line 188, in min_pit_process
    score_perms = torch.stack([loss[0] + loss[3],
Will finalize trainer extensions and updater before reraising the exception.
Traceback (most recent call last):
  File "/home/user/vyouman/espnet/egs/wsj_mix/asr1/../../../espnet/bin/asr_train.py", line 259, in <module>
    main(sys.argv[1:])
  File "/home/user/vyouman/espnet/egs/wsj_mix/asr1/../../../espnet/bin/asr_train.py", line 253, in main
    train(args)
  File "/home/user/vyouman/espnet/espnet/asr/pytorch_backend/asr_mix.py", line 323, in train
    trainer.run()
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/chainer/training/trainer.py", line 329, in run
    six.reraise(*sys.exc_info())
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/six.py", line 693, in reraise
    raise value
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/chainer/training/trainer.py", line 318, in run
    entry.extension(self)
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/chainer/training/extensions/evaluator.py", line 157, in __call__
    result = self.evaluate()
  File "/home/user/vyouman/espnet/espnet/asr/pytorch_backend/asr.py", line 104, in evaluate
    self.model(*x)
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 477, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py", line 123, in forward
    outputs = self.parallel_apply(replicas, inputs, kwargs)
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py", line 133, in parallel_apply
    return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/parallel/parallel_apply.py", line 77, in parallel_apply
    raise output
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/parallel/parallel_apply.py", line 53, in _worker
    output = module(*input, **kwargs)
  File "/home/user/vyouman/espnet/tools/venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 477, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/user/vyouman/espnet/espnet/nets/pytorch_backend/e2e_asr_mix.py", line 303, in forward
    wers.append(self.min_pit_process(tmp_wers) / len(sum(ref_words, [])))
  File "/home/user/vyouman/espnet/espnet/nets/pytorch_backend/e2e_asr_mix.py", line 188, in min_pit_process
    score_perms = torch.stack([loss[0] + loss[3],
IndexError: list index out of range

**Comment by vyouman** [comment]
Here is the configuration.

> # asr_train.py --ngpu 2 --backend pytorch --outdir exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.2_bs10_mli800_mlo150_seed1_lsmunigram0.05_delta/results --tensorboard-dir tensorboard/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.2_bs10_mli800_mlo150_seed1_lsmunigram0.05_delta --debugmode 1 --dict data/lang_1char/tr_units.txt --debugdir exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.2_bs10_mli800_mlo150_seed1_lsmunigram0.05_delta --minibatches 0 --verbose 1 --resume --seed 1 --train-json dump/tr/deltatrue/data.json --valid-json dump/cv/deltatrue/data.json --num-spkrs 2 --etype vggblstmp --elayers-sd 1 --elayers 2 --eunits 1024 --eprojs 1024 --subsample 1_2_2_1 --dlayers 1 --dunits 300 --atype location --adim 320 --awin 5 --aheads 4 --aconv-chans 10 --aconv-filts 100 --mtlalpha 0.2 --lsm-type unigram --lsm-weight 0.05 --batch-size 10 --maxlen-in 800 --maxlen-out 150 --sampling-probability 0.2 --opt adadelta --epochs 15 --patience 3 --beam-size 30 --spa-
# Started at Fri Mar  8 12:55:55 HKT 2019
#
2019-03-08 12:55:56,087 (asr_train:214) WARNING: CUDA_VISIBLE_DEVICES is not set.
2019-03-08 12:55:56,087 (asr_train:220) INFO: python path = (None)
2019-03-08 12:55:56,087 (asr_train:223) INFO: random seed = 1
2019-03-08 12:55:56,088 (asr_train:240) INFO: backend = pytorch
2019-03-08 12:55:56,939 (deterministic_utils:24) INFO: torch type check is disabled
2019-03-08 12:55:57,128 (asr_mix:130) INFO: #input dims : 249
2019-03-08 12:55:57,128 (asr_mix:131) INFO: #output dims: 50
2019-03-08 12:55:57,128 (asr_mix:142) INFO: Multitask learning mode
2019-03-08 12:55:57,129 (e2e_asr_mix:86) WARNING: Subsampling is not performed for vgg*. It is performed in max pooling layers at CNN.
2019-03-08 12:55:57,129 (e2e_asr_mix:87) INFO: subsample: 1 1 1 1
2019-03-08 12:55:57,129 (e2e_asr_mix:92) INFO: Use label smoothing with unigram
2019-03-08 12:56:00,153 (e2e_asr_mix:477) INFO: Use CNN-VGG + BLSTMP for encoder
2019-03-08 12:56:02,130 (asr_mix:161) INFO: writing a model config file to exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.2_bs10_mli800_mlo150_seed1_lsmunigram0.05_delta/results/model.json
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: aconv_chans: 10
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: aconv_filts: 100
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: adim: 320
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: aheads: 4
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: asr_model: False
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: atype: location
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: awin: 5
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: backend: pytorch
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: batch_size: 10
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: beam_size: 30
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: char_list: ['<blank>', '<unk>', '!', '"', '&', "'", '(', ')', '*', ',', '-', '.', '/', ':', ';', '<NOISE>', '<space>', '?', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '[', ']', '{', '}', '~', '<eos>']
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: criterion: acc
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: ctc_type: warpctc
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: ctc_weight: 0.3
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: debugdir: exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.2_bs10_mli800_mlo150_seed1_lsmunigram0.05_delta
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: debugmode: 1
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: dict: data/lang_1char/tr_units.txt
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: dlayers: 1
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: dropout_rate: 0.0
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: dropout_rate_decoder: 0.0
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: dtype: lstm
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: dunits: 300
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: early_stop_criterion: validation/main/acc
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: elayers: 2
2019-03-08 12:56:02,131 (asr_mix:164) INFO: ARGS: elayers_sd: 1
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: epochs: 15
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: eprojs: 1024
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: eps: 1e-08
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: eps_decay: 0.01
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: etype: vggblstmp
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: eunits: 1024
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: grad_clip: 5
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: lm_weight: 0.1
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: lsm_type: unigram
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: lsm_weight: 0.05
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: maxlen_in: 800
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: maxlen_out: 150
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: maxlenratio: 0.0
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: minibatches: 0
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: minlenratio: 0.0
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: mt_model: False
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: mtlalpha: 0.2
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: n_iter_processes: 0
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: nbest: 1
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: ngpu: 2
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: num_save_attention: 3
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: num_spkrs: 2
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: opt: adadelta
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: outdir: exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.2_bs10_mli800_mlo150_seed1_lsmunigram0.05_delta/results
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: patience: 3
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: penalty: 0.0
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: preprocess_conf: None
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: report_cer: True
2019-03-08 12:56:02,132 (asr_mix:164) INFO: ARGS: report_wer: True
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: resume: None
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: rnnlm: None
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: rnnlm_conf: None
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: sampling_probability: 0.2
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: seed: 1
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: sortagrad: 0
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: spa: True
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: subsample: 1_2_2_1
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: sym_blank: <blank>
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: sym_space: <space>
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: tensorboard_dir: tensorboard/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.2_bs10_mli800_mlo150_seed1_lsmunigram0.05_delta
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: threshold: 0.0001
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: train_json: dump/tr/deltatrue/data.json
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: valid_json: dump/cv/deltatrue/data.json
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: verbose: 1
2019-03-08 12:56:02,133 (asr_mix:164) INFO: ARGS: weight_decay: 0.0
2019-03-08 12:56:02,145 (asr_mix:172) INFO: batch size is automatically increased (10 -> 20)
2019-03-08 12:56:05,381 (asr_mix_utils:43) INFO: # utts: 20000
2019-03-08 12:56:05,387 (asr_mix_utils:82) INFO: # minibatches: 1517
2019-03-08 12:56:05,391 (asr_mix_utils:43) INFO: # utts: 5000
2019-03-08 12:56:05,462 (asr_mix_utils:82) INFO: # minibatches: 387

**Comment by vyouman** [comment]
There are 1517 minibatches. I checked the training log. It seems OK with the first 1500 minibatches, but cracks almost at the end of an epoch.

This is part of the log in the exp/tr_pytorch_vggblstmp_sde1_rece2_subsample1_2_2_1_unit1024_proj1024_d1_unit300_location_spatrue_aconvc10_aconvf100_mtlalpha0.2_adadelta_sampprob0.2_bs10_mli800_mlo150_seed1_lsmunigram0.05_delta/results/log. 

> [
> ......,
>     {   
>         "main/loss_ctc": 267.8714545440674,
>         "main/loss_att": 228.00763526916504,
>         "main/acc": 0.43536497100482696,
>         "main/cer": 0.0,
>         "main/wer": 0.0,
>         "main/loss": 235.98040367126464,
>         "eps": 1e-08,
>         "epoch": 0,
>         "iteration": 1300,
>         "elapsed_time": 2840.377956914017
>     },  
>     {   
>         "main/loss_ctc": 257.64139945983885,
>         "main/loss_att": 223.73684577941896,
>         "main/acc": 0.44499859636888367,
>         "main/cer": 0.0,
>         "main/wer": 0.0,
>         "main/loss": 230.517759475708,
>         "eps": 1e-08,
>         "epoch": 0,
>         "iteration": 1400,
>         "elapsed_time": 3053.219671224011
>     },  
>     {   
>         "main/loss_ctc": 253.85887420654296,
>         "main/loss_att": 218.81052642822266,
>         "main/acc": 0.4567279678813855,
>         "main/cer": 0.0,
>         "main/wer": 0.0,
>         "main/loss": 225.82020050048828,
>         "eps": 1e-08,
>         "epoch": 0,
>         "iteration": 1500,
>         "elapsed_time": 3264.480705066002
>     } 
> ]

**Comment by sw005320** [comment]
Thanks for the report. We did not test in such a case, and there may be some issues. I'm also thinking that some wrong behaviors may happen with multiple GPUs. @simpleoier, could you also test it with your environment by making report_cer and report_wer True?

**Comment by simpleoier** [comment]
It's my fault. I didn't test report_cer and report_wer for multi-speaker case. The ``PIT'' computation of cer and wer was incomplete. I already create a PR to correct it.

**Comment by sw005320** [comment]
Thanks @simpleoier for your quick action!
@vyouman, could you test it with the new version?

**Comment by vyouman** [comment]
@sw005320 @simpleoier I tested with the new version with 2 GPUs and set  2 minibatches (N) to use for debugging with report_cer and report_wer set to be True. The program is so slow that I suspect that if it is working. Maybe it's not a good choice to set report_cer and report_wer in the training procedure.  It is very time-consuming. I started the program at 11:44 am but the program is still running now (14:35 pm).
